Classes/FairPlayDRM/FPSContentKeySessionDelegate.swift

```diff
@@ -0,0 +1,118 @@
+class FPSContentKeySessionDelegate: NSObject, AVContentKeySessionDelegate {
+        
+    var assetHelpersMap = [String: FPSLicenseHelper]()
+    
+    func contentKeySession(_ session: AVContentKeySession, didProvide keyRequest: AVContentKeyRequest) {
+        try? handleContentKeyRequest(keyRequest: keyRequest)
+    }
+    
+    func contentKeySession(_ session: AVContentKeySession, didProvide keyRequest: AVPersistableContentKeyRequest) {
+        try? handleContentKeyRequest(keyRequest: keyRequest)
+    }
+
+    func contentKeySession(_ session: AVContentKeySession, didProvideRenewingContentKeyRequest keyRequest: AVContentKeyRequest) {
+        try? handleContentKeyRequest(keyRequest: keyRequest)
+    }
+    
+    func contentKeySession(_ session: AVContentKeySession,
+                           didUpdatePersistableContentKey persistableContentKey: Data,
+                           forContentKeyIdentifier keyIdentifier: Any) {
+        
+        #if DEBUG
+        fatalError("Dual Expiry feature not implemented")
+        #endif
+    }
+
+    func contentKeySession(_ session: AVContentKeySession, shouldRetry keyRequest: AVContentKeyRequest,
+                           reason retryReason: AVContentKeyRequestRetryReason) -> Bool {
+        
+        var shouldRetry = false
+        
+        switch retryReason {
+            /*
+             Indicates that the content key request should be retried because the key response was not set soon enough either
+             due the initial request/response was taking too long, or a lease was expiring in the meantime.
+             */
+        case AVContentKeyRequestRetryReason.timedOut:
+            shouldRetry = true
+            
+            /*
+             Indicates that the content key request should be retried because a key response with expired lease was set on the
+             previous content key request.
+             */
+        case AVContentKeyRequestRetryReason.receivedResponseWithExpiredLease:
+            shouldRetry = true
+            
+            /*
+             Indicates that the content key request should be retried because an obsolete key response was set on the previous
+             content key request.
+             */
+        case AVContentKeyRequestRetryReason.receivedObsoleteContentKey:
+            shouldRetry = true
+            
+        default:
+            break
+        }
+        
+        return shouldRetry
+    }
+    
+    // Informs the receiver a content key request has failed.
+    func contentKeySession(_ session: AVContentKeySession, contentKeyRequest keyRequest: AVContentKeyRequest, didFailWithError err: Error) {
+        // Add your code here to handle errors.
+    }
+    
+    func assetHelper(_ keyIdentifier: Any?) -> FPSLicenseHelper? {
+        guard let id = keyIdentifier as? String else { return nil }
+        return assetHelpersMap[id]
+    }
+    
+    func handleContentKeyRequest(keyRequest: AVContentKeyRequest) throws {
```

Comment:
Why is throws needed? I didn't see it throws anything.

Reply:
Leftovers

Reply:
Leave it, for the addition to the below fix for iOS 11.2